example/router/detail/router.hpp

```diff
@@ -0,0 +1,256 @@
+        // {id+}
+        plus
+    };
+
+    std::string str_;
+    bool is_literal_ = true;
+    modifier modifier_ = modifier::none;
+
+    friend struct segment_template_rule_t;
+public:
+    segment_template() = default;
+
+    BOOST_URL_DECL
+    bool
+    match(pct_string_view seg) const;
+
+    string_view
+    string() const
+    {
+        return str_;
+    }
+
+    BOOST_URL_DECL
+    string_view
+    id() const;
+
+    bool
+    empty() const
+    {
+        return str_.empty();
+    }
+
+    bool
+    is_literal() const
+    {
+        return is_literal_;
+    }
+
+    bool
+    has_modifier() const
+    {
+        return !is_literal() &&
+               modifier_ != modifier::none;
+    }
+
+    bool
+    is_optional() const
+    {
+        return modifier_ == modifier::optional;
+    }
+
+    bool
+    is_star() const
+    {
+        return modifier_ == modifier::star;
+    }
+
+    bool
+    is_plus() const
+    {
+        return modifier_ == modifier::plus;
+    }
+
+    friend
+    bool operator==(
+        segment_template const& a,
+        segment_template const& b)
+    {
+        if (a.is_literal_ != b.is_literal_)
+            return false;
+        if (a.is_literal_)
+            return a.str_ == b.str_;
+        return a.modifier_ == b.modifier_;
+    }
+
+    // segments have precedence:
+    //     - literal
+    //     - unique
+    //     - optional
+    //     - plus
+    //     - star
+    friend
+    bool operator<(
+        segment_template const& a,
+        segment_template const& b)
+    {
+        if (b.is_literal())
+            return false;
+        if (a.is_literal())
+            return !b.is_literal();
+        return a.modifier_ < b.modifier_;
+    }
+};
+
+// A segment template is either a literal string
+// or a replacement field (as in a format_string).
+// Fields cannot contain format specs and might
+// have one of the following modifiers:
+// - ?: optional segment
+// - *: zero or more segments
+// - +: one or more segments
+struct segment_template_rule_t
+{
+    using value_type = segment_template;
+
+    BOOST_URL_DECL
+    urls::result<value_type>
+    parse(
+        char const*& it,
+        char const* end
+    ) const noexcept;
+};
+
+constexpr auto segment_template_rule = segment_template_rule_t{};
+
+constexpr auto path_template_rule =
+    grammar::tuple_rule(
+        grammar::squelch(
+            grammar::optional_rule(
+                grammar::delim_rule('/'))),
+        grammar::range_rule(
+            segment_template_rule,
+            grammar::tuple_rule(
+                grammar::squelch(grammar::delim_rule('/')),
+                segment_template_rule)));
+
+// a small vector for child nodes...
+// we shouldn't expect many children per node, and
+// we don't want to allocate for that. But we also
+// cannot cap the max number of child nodes because
+// especially the root nodes can potentially an
+// exponentially higher number of child nodes.
+class child_idx_vector
```

Comment:
without benchmarks, this is kind of superfluous

Reply:
Yes. Cache locality more than allocations. But still, yes.